src/helpers/upload.ts

```diff
@@ -0,0 +1,129 @@
+import retry from 'async-retry'
+import {AxiosPromise, AxiosRequestConfig, AxiosResponse} from 'axios'
+import FormData from 'form-data'
+import {ReadStream} from 'fs'
+
+import {ApiKeyValidator} from './apikey'
+import {RequestBuilder} from './interfaces'
+
+const errorCodesNoRetry = [400, 403, 413]
+
+/** Multipart payload destined to be sent to Datadog's API
+ */
+export interface MultipartPayload {
+  content: Map<string, MultipartValue>
+}
+
+export interface MultipartValue {
+  options?: FormData.AppendOptions | string
+  value: string | ReadStream
+}
+
+export interface UploadOptions {
+  /** ApiKeyValidator (optional) throws an InvalidConfiguratonException when upload fails because
+   * of an invalid API key. Callers should most likely catch this exception and display it as a
+   * nice error message.
+   */
+  apiKeyValidator?: ApiKeyValidator
+
+  /** Retries is the amount of upload retries before giving up. Some requests are never retried
+   * (400, 413).
+   */
+  retries: number
+
+  /** Callback when upload fails (retries are not considered as failure)
+   */
+  onError(error: Error): void
+
+  /** Callback to execute before retries
+   */
+  onRetry(e: Error, attempts: number): void
```

Comment:
Nit: to be consistent with other callback definition
```suggestion
  onRetry(error: Error, attempts: number): void
```